README.md
# SCIRA-package
Single Cell Inference of Regulatory Activity

`SCIRA` is such a hybrid approach which uses the inferred predicted targets (or "regulon") of a given transcription factor from bulk data to infer its regulatory activity in each single cell. It encompasses two main steps:

1. Construction of a tissue-specific transcription factor regulatory network, consisting of transcription factors that are more highly expressed in the user-specified tissue type (the 'tissue type of interest') compared to other tissue types, plus an associated set of high-confidence downstream targets.
2. Estimation of transcription factor activity in this network, in any given single cell of a scRNA-seq dataset.


## Usage
#### Inferring tissue-specific network

```{r eval=FALSE}
net.o <- sciraInfNet(data=data.m, tissue=colnames(data.m), toi = "Lung", cft = c("Blood","Spleen"),
         TFs = TFeid, sdth = 0.25, sigth = 0.05, pcorth = 0.2, degth = c(0.05, 0.05),
         lfcth = c(log2(1.5), 0), minNtgts = 5, ncores = 1)
```
** Note: `data.m` must be an intra-sample log-normalized bulk-tissue mRNA expression (RNA-Seq) dataset, like the dataset from GTEX. 

#### Estimating transcription factor activity
```{r eval=FALSE}
TFact <- sciraRegAct(data = data.m, regnet = net.o$netTOI, norm = "z", ncores = 1)
```
#### Installation
An easy way to install SCIRA is by facilitating the devtools R package.
```{r eval=FALSE}
#install.packages("devtools")
library(devtools)
install_github("WangNing0420/SCIRA", build_opts = c("--no-resave-data", "--no-manual"))
```

## Getting started
The SCIRA package contains a tutorial showing people how to implement SCIRA in their work. The tutorial can be found in the package-vignette:

```{r eval=FALSE}
library(SCIRA)

vignette("SCIRA")
```

## References

Chen Y, Widschwendter M, and Teschendorff AE. 2017. “Systems-Epigenomics Inference of Transcription Factor Activity Implicates Aryl-Hydrocarbon-Receptor Inactivation as a Key Event in Lung Cancer Development.” Genome Biol 18:236.

Wang, N., & Teschendorff, AE. 2019.  “Leveraging high-powered RNA-Seq datasets to improve inference of regulatory activity in single-cell RNA-Seq data.“ BioRxiv, 553040. 


NEWS
This file provides information on package updates.


R/LimmaFn.R
#' @title Do differential gene expression
#'
#' @description \code{LimmaFn} uses functions in \code{limma} package to easily compute the moderated t-statistics and p-values from differential gene expression tests comparing between different phenotypes even when sample size is small.
#'
#' @param pheno A vector of sample phenotypes. Sample phenotype in a scientific research could be treatment/control, normal/cancer or smoker/non-smoker. Different phenotypes could each be encoded as 0/1 when inputting to \code{LimmaFn}, for example, Normal-0; Cancer-1.
#' @param data A matrix, the unnormalized gene expression dataset, should be a numeric matrix, with rows referring to genes and columns to samples. In this matrix you could use any type of gene IDs, like Entrez ID, Ensembl ID, HUGO gene symbol... But make sure to use the same gene annotation through out your analysis.
#'
#' @return A table with rows for all genes (ranked by significance) and columns of log2 fold-change, average expression, moderated t-statistic, p-value, adjusted p-value (Benjamini–Hochberg procedure). The table is the output of \code{\link[limma]{topTable}} function.
#'
#' @details This function computes the moderated t-statistic for users using empirical Bayes method, it is especially useful when the sample size is too small to perform parametric tests.
#'
#' Given a gene expression data matrix and a vector indicating sample phenotype, \code{LimmaFn} first fits a linear model using \code{\link[limma]{lmFit}}, then it refits the model and do comparisons between any two different phenotypes with \code{\link[limma]{contrasts.fit}}, finally it estimates moderated t-statistics for each comparison from the fitted model using empirical Bayes method (\code{\link[limma]{eBayes}}) and output the result from the \code{\link[limma]{topTable}} function.
#'
#' Note that doing the \code{contrasts.fit} step will not make a difference if you do comparison between two different sample status (treatment/control). However, When there are more than two sample status in your data set, this step will do comparison between every two status. And resulted summary tables will be stored in a list.
#'
#' @seealso \code{\link[limma]{lmFit}} for fitting a linear model, \code{\link[limma]{contrasts.fit}} for refitting, \code{\link[limma]{eBayes}} for Bayes method, \code{\link[limma]{topTable}} for the output table.
#'
#' @importFrom stats cor
#' @importFrom stats lm
#' @importFrom stats model.matrix
#' @importFrom stats pbinom
#' @importFrom stats pnorm
#' @importFrom stats sd
#'
#' @export LimmaFn
#'
#' @examples
#' # prepare the phenotype info ("C"-control; "T"-treatment)
#' pheno.v <- c("C","C","C","T","T","T")
#' # prepare your normalized data matrix.
#' data.m <- matrix(rnorm(120),nrow=20,ncol=6)
#' # run function
#' lim.o <- LimmaFn(pheno.v, data.m)

LimmaFn <- function(pheno.v,data.m){
  
  
  
  ### construct model matrix
  
  sampletype.f <- as.factor(pheno.v);
  
  design.sample <- model.matrix(~0 + sampletype.f);
  
  colnames(design.sample) <- levels(sampletype.f);
  
  sampletypes.v <- levels(sampletype.f);
  
  
  
  ### do linear model fit
  
  lmf.o <- lmFit(data.m,design.sample);
  
  
  
  ### construct contrast matrix
  
  ntypes <- length(levels(sampletype.f));
  
  ncomp <- 0.5*ntypes*(ntypes-1);
  
  cont.m <- matrix(0,nrow=ncol(design.sample),ncol=ncomp);
  
  tmp.v <- vector();
  
  c <- 1;
  
  for(i1 in 1:(ntypes-1)){
    
    for(i2 in (i1+1):ntypes){
      
      cont.m[i1,c] <- -1;
      
      cont.m[i2,c] <- 1;
      
      tmp.v[c] <- paste(sampletypes.v[i2],"--",sampletypes.v[i1],sep="");
      
      c <- c+1;
      
    }
    
  }
  
  rownames(cont.m) <- sampletypes.v; # sampletype.v determined separately
  
  colnames(cont.m) <- tmp.v;
  
  
  
  ### do linear model to contrasts
  
  lmf2.o <- contrasts.fit(lmf.o,cont.m);
  
  
  
  ### empirical Bayesian estimation of differentially expressed genes (DEGs)
  
  bay.o <- eBayes(lmf2.o);
  
  
  
  ### build ranked list of DEGs for each comparison
  
  top.lm <- list();
  
  for(c in 1:ncol(cont.m)){
    
    top.lm[[c]] <- topTable(bay.o,coef=c,adjust="fdr",number=nrow(data.m));
    
  }
  
  
  
  return(list(top=top.lm,cont=cont.m));
  
  
  
} ## end of limma function

R/InferTFact.R
#' @title Estimate TF activity score in a sample
#'
#' @description \code{InferTFact} is an auxiliary function for function \code{sciraRegAct}.
#'
#' @param exp A numeric vector of gene expression levels for all TF target genes in a sample.
#' @param regnet A matrix, the network estimated by function \code{sciraInfNet}, with +1 referring to positive regulation, -1 negative regulation, and 0 no regulation.
#'
#' @return A vector storing the activity scores of all TFs in a specific sample.
#'
#' @details \code{InferTFact} regresses the expression profiles of TF target genes against the binding profile of this TF on these genes. The output t-statistics are taken as the TF activity scores.
#'
#' @importFrom stats lm
#'
#' @examples
#' # TF regulatory network with 100 genes and 5 TFs.
#' regnet.m <- matrix(sample(c(-1, 0, 1), 500, replace = TRUE), nrow = 100)
#' # gene expression vector (for one sample)
#' exp.v <- rnorm(100)
#' # TF activity score
#' TFact.v <- SCIRA:::InferTFact(exp.v, regnet.m)

InferTFact <- function(exp, regnet) {
  act.v <- apply(regnet, 2, function(tmp.v) {
    lm.o <- lm(exp ~ tmp.v)
    act <- summary(lm.o)$coeff[2, 3]
    return(act)
  })

  return(act.v)
}

#' @title Estimate TF activity score in different samples
#'
#' @description \code{InferTFactPRL } is an auxiliary function for function \code{sciraRegAct}.
#'
#' @param idx A numeric vector indicating the samples from which you want to estimate the TF activity score.
#' @param tmp A numeric matrix of gene expression levels for all TF target genes, with rows referring to genes, columns to samples.
#' @param regnet The network estimated from function \code{sciraInfNet}, with +1 referring to positive regulation, -1 negative regulation, and 0 no regulation.
#'
#' @return A vector storing the activity scores of all TFs from user specified sample in the data matrix.
#'
#' @details The user needs to input the index of the sample in the data matrix from which \code{InferTFactPRL} estimates TF activity. \code{InferTFactPRL} accomplishes the work by regressing the expression profile of TF target genes against the binding profile of this TF on these genes. The output t-statistics are taken as the TF activity score.
#'
#' @importFrom stats lm
#'
#' @examples
#' # TF regulatory network with 100 genes and 5 TFs.
#' regnet.m <- matrix(sample(c(-1, 0, 1), 500, replace = TRUE), nrow = 100)
#' # gene expression data matrix with 100 genes and 10 samples.
#' tmp.m <- matrix(rnorm(1000), nrow = 100)
#' # TF activity score
#' TFact.v <- SCIRA:::InferTFactPRL(1, tmp.m, regnet.m)

InferTFactPRL <- function(idx, tmp, regnet) {
  exp.v <- tmp[, idx]
  act.v <- apply(regnet, 2, function(tmp.v) {
    lm.o <- lm(exp.v ~ tmp.v)
    act <- summary(lm.o)$coeff[2, 3]
    return(act)
  })

  return(act.v)
}


R/ComputePCOR.R
#' @title Compute partial correlation coefficient
#'
#' @description \code{ComputePCOR} computes partial correlation between transcription regulators and their targets.
#'
#' @param idx Numeric, the index of target genes in the row of a binarized network \code{selbinNET.m}.
#' @param mapTG.idx A vector of indexes output from \code{match} function when mapping the target genes in the network to rows of the original expression data matrix.
#' @param mapTF.idx A vector of indexes output from \code{match} function when mapping the transcription factors (TFs) in the network to rows of the original expression data matrix.
#' @param selbinNET.m A matrix, the binarized network with rows referring to TF target genes, and columns to TFs (the regulators). 0s means no regulation between TF-gene, while 1s means significant regulation (either positive or negative).
#' @param exp A matrix, the original gene expression data matrix across different tissue types with rows referring to genes, columns to samples.
#'
#' @return A matrix with partial correlation coefficients between TF targets and their regulators.
#'
#' @details This function is used to calculate partial correlations between a TF target gene and all its regulators.
#'
#' User needs to first provide the index (\code{idx}) of the gene in the binarized matrix, then all its regulators are detected from the network. Then the expression profile of this gene and its regulators are retrieved from the expression data matrix (\code{exp}) with \code{mapTG.idx} and \code{mapTF.idx} indexes which indicates the number of rows they are in \code{exp} (from \code{match} function).
#'
#' The output could be:
#' 1) a matrix storing the partial correlation coefficients;
#' 2) NULL if the selected gene has only one regulator according to the binarized network.
#'
#' @importFrom stats cor

ComputePCOR <- function(idx,mapTG.idx,mapTF.idx,selbinNET.m,exp.m){
  g <- idx
  reg.idx <- which(selbinNET.m[g,]==1)
  if(length(reg.idx)>=2){
    tmp.idx <- c(mapTG.idx[g],mapTF.idx[reg.idx])
    cor.m <- cor(t(exp.m[tmp.idx,]))
    pcor.m <- cor2pcor(cor.m)
  }
  else{
    pcor.m <- NULL
  }
  return(pcor.m)
}

R/sciraRegAct.R
#' @title Infer TF activity from single cell gene expression profile
#'
#' @description \code{sciraRegAct} calculates TF activity scores in user input data set. It could be a single cell gene expression dataset.
#'
#' @param data A gene expression data matrix, with rows referring to genes and columns to samples.
#' @param regnet A matrix, the regulatory network inferred from \code{sciraInfNet} function.
#' @param norm A character indicating the method used to normalize your input data set, "c" for "centering"; "z" for "z-score normalization".
#' @param ncores A numeric, the number of cores to use. See \code{mclapply}.
#'
#' @return A matrix of TF activity score with rows referring to TFs, columns to samples.
#'
#' @details \code{sciraRegAct} is one of the two main functions in \code{SCIRA} package. It takes the output regulatory network from \code{sciraInfNet} as input, and computes the activity of all TFs in this network from user provided \code{data}.
#'
#' The \code{data} matrix should be single cell gene expression data, with rows are genes and columns are samples. Duplicated row names are not allowed, so you should average the these rows before running \code{sciraRegAct}.
#'
#' Note that it's very important that you use the same gene identifier through out the whole analysis.
#'
#' @import parallel
#' @importFrom stats cor
#' @importFrom stats lm
#' @importFrom stats model.matrix
#' @importFrom stats pbinom
#' @importFrom stats pnorm
#' @importFrom stats sd
#'
#' @export sciraRegAct
#'
#' @examples
#' # estimate TF activity
#' # single cell RNA-seq dataset
#' data("scLiver")
#' # Lungnet
#' data("netLiver")
#' TFact <- sciraRegAct(scLiver,regnet=netLiver,norm="z",ncores=1)

sciraRegAct <- function(data, regnet, norm = c("c", "z"), ncores = 4) {


  if (is.vector(data)) {
    inter <- intersect(names(data), rownames(regnet))
    data <- data[inter]
    regnet <- regnet[inter, ]
    actTF <- InferTFact(data, regnet)
    names(actTF) <- colnames(regnet)
  }
  else if (is.matrix(data)) {
    common.v <- intersect(rownames(data),rownames(regnet));
    match(common.v,rownames(data)) -> map1.idx;
    match(common.v,rownames(regnet)) -> map2.idx;        
    ndata <- data[map1.idx,] - rowMeans(data[map1.idx,]);
    
    if(norm=="z"){
      sd.v <- apply(data[map1.idx,],1,sd);
      nz.idx <- which(sd.v>0);
      z.idx <- which(sd.v==0);
      ndata <- data[map1.idx,];       
      ndata[nz.idx,] <- (data[map1.idx[nz.idx],] - rowMeans(data[map1.idx[nz.idx],]))/sd.v[nz.idx];
      ndata[z.idx,] <- 0;
    }
    idx.l <- as.list(1:ncol(data));
    prl.o <- mclapply(idx.l,InferTFactPRL,ndata,regnet[map2.idx,],mc.cores=ncores);
    actTF <- matrix(unlist(prl.o), nrow = ncol(regnet), ncol=  length(prl.o), byrow = FALSE)
    rownames(actTF) <- colnames(regnet)
    colnames(actTF) <- colnames(data)
  }

  return(actTF)
}


R/sepiraRegAct.R
#' @title Infer TF activity from single cell gene expression profile
#'
#' @description \code{sciraRegAct} calculates TF activity scores in user input data set. It could be a single cell gene expression dataset.
#'
#' @param data A gene expression data matrix, with rows referring to genes and columns to samples.
#' @param regnet A matrix, the regulatory network inferred from \code{sciraInfNet} function.
#' @param norm A character indicating the method used to normalize your input data set, "c" for "centering"; "z" for "z-score normalization".
#' @param ncores A numeric, the number of cores to use. See \code{mclapply}.
#'
#' @return A matrix of TF activity score with rows referring to TFs, columns to samples.
#'
#' @details \code{sciraRegAct} is one of the two main functions in \code{SCIRA} package. It takes the output regulatory network from \code{sciraInfNet} as input, and computes the activity of all TFs in this network from user provided \code{data}.
#'
#' The \code{data} matrix should be single cell gene expression data, with rows are genes and columns are samples. Duplicated row names are not allowed, so you should average the these rows before running \code{sciraRegAct}.
#'
#' Note that it's very important that you use the same gene identifier through out the whole analysis.
#'
#' @import parallel
#' @importFrom stats cor
#' @importFrom stats lm
#' @importFrom stats model.matrix
#' @importFrom stats pbinom
#' @importFrom stats pnorm
#' @importFrom stats sd
#'
#' @export sciraRegAct
#'
#' @examples
#' # estimate TF activity
#' data("scLiver")
#' data("netLiver")
#' TFact <- sciraRegAct(scLiver,regnet=netLiver,norm="z",ncores=1)

sciraRegAct <- function(data, regnet, norm = c("c", "z"), ncores = 4) {


  if (is.vector(data)) {
    inter <- intersect(names(data), rownames(regnet))
    data <- data[inter]
    regnet <- regnet[inter, ]
    actTF <- InferTFact(data, regnet)
    names(actTF) <- colnames(regnet)
  }
  else if (is.matrix(data)) {
    common.v <- intersect(rownames(data),rownames(regnet));
    match(common.v,rownames(data)) -> map1.idx;
    match(common.v,rownames(regnet)) -> map2.idx;        
    ndata <- data[map1.idx,] - rowMeans(data[map1.idx,]);
    
    if(norm=="z"){
      sd.v <- apply(data[map1.idx,],1,sd);
      nz.idx <- which(sd.v>0);
      z.idx <- which(sd.v==0);
      ndata <- data[map1.idx,];       
      ndata[nz.idx,] <- (data[map1.idx[nz.idx],] - rowMeans(data[map1.idx[nz.idx],]))/sd.v[nz.idx];
      ndata[z.idx,] <- 0;
    }
    idx.l <- as.list(1:ncol(data));
    prl.o <- mclapply(idx.l,InferTFactPRL,ndata,regnet[map2.idx,],mc.cores=ncores);
    actTF <- matrix(unlist(prl.o), nrow = ncol(regnet), ncol=  length(prl.o), byrow = FALSE)
    rownames(actTF) <- colnames(regnet)
    colnames(actTF) <- colnames(data)
  }

  return(actTF)
}


R/sciraInfNet.R
#' @title Infer tissue-specific regulatory network from gene expression data
#'
#' @description \code{sciraInfNet()} is one of the two main functions in package \code{SCIRA}. Using it you can estimate tissue-specific regulatory networks in any tissue type of interest.
#'
#' @param data A matrix, the gene expression data matrix, with rows referring to unique genes and columns to samples from different tissue types.
#' @param tissue A phenotype vector, indicating the tissue types of samples. It should have the same order as the columns of the matrix.
#' @param toi A character, the tissue type of interest, a character telling the function the tissue type for which a user wants to estimate the network.
#' @param cft A character or character vector containing the tissues which are deemed to be confounding. It can be blood and/or spleen, which we found using \code{ESTIMATE} package that they contain extremely high proportion of immune and stromal cells. If specified, must be one of the tissues in \code{tissve}
#' @param TFs A vector of TFs. Note that one should use the same annotation in different data sets throughout the analysis. e.g. a vector of human transcription factors. However, the regulators need not be transcription factors.
#' @param sdth A numeric, the standard deviation threshold used to remove genes with little or zero standard deviation of its expression levels.
#' @param sigth A numeric, the unadjusted p-value threshold used to call significant interactions after calculating the correlation coefficients between TFs and target genes. This threshold is used to binarize the correlation coefficient matrix. If this value is not specified by user, the function will do Bonferroni correction and then use 0.05 as the threshold.
#' @param pcorth A numeric, the partial correlation threshold, in the range between 0 and 1, used to remove indirect interactions between TFs and their target genes.
#' @param degth A vector of length two, thresholds of adjusted p-value to call significant TFs in 1) comparison between \code{toi} and all other tissue types; 2)  comparison between \code{toi} and blood/spleen in \code{cft}.
#' @param lfcth A vector of length two, thresholds of log2(fold-change) to call significant TFs in 1) comparison between \code{toi} and all other tissue types; 2)  comparison between \code{toi} and blood/spleen in \code{cft}.
#' @param minNtgts An integer used to filter out TFs with few targets. Only TFs with more than `minNtgts` target genes can be kept in the network.
#' @param ncores An integer, the number of cores to use when computing partial correlation. See \code{mclapply}.
#'
#' @return A list with three entries:
#'
#' \code{$netTOI} the tissue specific network, rows refer to TF target genes, while columns refer to TFs. Entries are 1 for activating interactions, 0 no interactions, and -1 for inhibiting interactions.
#'
#' \code{$sumnet} a matrix summarizing the number of TF target genes and the number of positively/negatively regulated target genes for each TF in the inferred network.
#'
#' \code{$top} a list, entries are the tables summarizing the results of differential expression analyses. The first is the table from comparison between \code{toi} and 1) all other tissue; The rest tables are resulted from comparison to 2) the blood or/ and 3) spleen.
#'
#' @details \code{sciraInfNet} generates tissue specific TF regulatory networks from gene expression data across multi-tissue types.
#'
#' The gene expression data set \code{data} should not be normalized by user before inputting to \code{sciraInfNet}, with rows referring to genes and columns to samples from different tissue types. Duplicated gene names/IDs should be averaged.
#'
#' The user needs to input the tissue type of samples (\code{tissue}) in the data set as well as the tissue type of interest (\code{toi}). Please make sure the \code{toi} is in the \code{tissue} and spelled correctly.
#'
#' Using differential gene expression analysis, we detect TFs that are highly active in \code{toi} and less active in other tissue types. When doing such analyses, the results could be confounded due to cell-type heterogeneity. \code{sciraInfNet} provides a way to adjust for immune/ stromal cell contamination by doing additional comparisons between \code{toi} and 1) blood; 2)spleen as long as expression data for any one/ both of the tissue types are available in \code{data}.
#'
#' \code{TFs} is a vector containing the identifiers of all TFs (regulators). In our paper we used the 1385 TFs annotated as "transcription factors" in MSigDB. You could input your own list of TFs to \code{sciraInfNet}.
#'
#' \code{sdth} is a standard deviation threshold that is used to remove genes in user provided data set which are with small or close to zero standard deviation. By default the threshold is 0.25.
#'
#' From the gene expression data matrix \code{sciraInfNet} estimates Pearson correlation coefficient between every TF-gene pair as well as corresponding p-value. The p-value threshold \code{sigth} binarizes the network into "regulation" (1) /"no regulation" (0). This binarized network is used to determine the covariants when estimating the partial correlation between target genes and their regulators (TFs).
#'
#' \code{pcorth} is the partial correlation coefficient threshold for calling significant direct TF-gene interactions. By default \code{pcorth} equals 0.2.
#'
#' \code{degth} and \code{lfcth} are vectors each contains the 3 thresholds for adjusted p-value/log2 fold-change to call significant TFs in comparisons between \code{toi} and 1) all other tissue types; 2) the 1st tissue type (blood) in \code{cft}; 3) the 2nd tissue type (spleen or adipose tissue) in \code{cft}. These differential expression analyses are done to find tissue-specific TFs that are only highly activated in tissue type of interest.
#'
#' When having detected tissue-specific TFs, we could get a network with only these TFs and their target genes. However \code{sciraInfNet} further remove TFs with less than \code{minNtgts} target genes. By default the minimal number of TF targets in the final network is 10.
#'
#' The step of calculating partial correlation coefficients is done by in parallel, by default \code{sciraInfNet} splits the work into 4 sub-processes. User could use more cores by specifying parameter \code{ncores}.
#'
#' @import parallel
#' @import corpcor
#' @import limma
#'
#' @export sciraInfNet
#'
#' @examples
#' # gene expression data set (a subset of GTEx data set)
#' data("GeneExp")
#' # TFs
#' data("TFeid")
#' # run the function
#' # The resulted network is small due to the limited size of the 'GeneExp' data set
#' net.o <- sciraInfNet(GeneExp,colnames(GeneExp), toi="Lung", cft=c("Blood","Spleen"), TFs=TFeid, sdth=0.25, sigth=0.05, pcorth=0.2, degth=rep(0.03,2), lfcth=c(log2(1.5),0), minNtgts=3, ncores=1)
#' # degth = c(0.05, 0.05)' is recommended
#' 
sciraInfNet <- function(data, tissue, toi, cft = NULL, TFs, sdth = 0.25, sigth = NULL, pcorth = 0.2, degth = c(0.05, 0.05), lfcth = c(1, log2(1.5)), minNtgts = 10, ncores = 4) {

  tt.v <- levels(factor(tissue))
  
  if (length(intersect(toi,tt.v)) == 0) stop("Your tissue of interest is not in 'tissue', or you have mispelled toi")
  if (FALSE %in% (cft %in% tt.v)) stop(paste0("'", cft[!cft %in% tt.v], "'", " is not in 'tissue', or you have mispelled it."))

  ### remove genes with no or little variance
  sd.v <- apply(data, 1, sd)
  selG.idx <- which(sd.v > sdth)
  exp.m <- data[selG.idx, ]
  
  ### find representation of regulators in data, and define regulatees/targets
  tfEID.v <- TFs
  repTF.v <- intersect(tfEID.v, rownames(exp.m))
  tgtsEID.v <- setdiff(rownames(exp.m), tfEID.v)
  mapTF.idx <- match(repTF.v, rownames(exp.m))
  mapTGTS.idx <- match(tgtsEID.v, rownames(exp.m))
  
  ### compute correlations and estimate P-values
  corNET.m <- cor(t(exp.m[mapTGTS.idx, ]), t(exp.m[mapTF.idx, ]))
  zNET.m <- 0.5 * log( (1 + corNET.m) / (1 - corNET.m))
  stdev <- 1/sqrt(ncol(exp.m)-3) 
  pvNET.m <- 2 * pnorm(abs(zNET.m), 0, stdev, lower.tail = FALSE)
  
  
  ### for each gene, now identify the TFs which are correlated univariately- for these then run multivariate regression
  if (is.null(sigth)) {
    sigth <- 0.05 / prod(dim(pvNET.m))
  }
  binNET.m <- pvNET.m
  binNET.m[pvNET.m < sigth] <- 1
  binNET.m[pvNET.m >= sigth] <- 0
  if(sum(binNET.m) > 0.01 * prod(dim(pvNET.m))){### if more than 1% cap at 1%
    topE <- floor(0.01 * prod(dim(pvNET.m)))

    zNET.v <- as.vector(abs(zNET.m))
    tmp.s <- sort(zNET.v, decreasing = TRUE, index.return = TRUE)
    out.v <- rep(0, length(zNET.v))
    out.v[tmp.s$ix[seq_len(topE)]] <- 1
    
    binNET.m <- matrix(out.v,nrow = nrow(zNET.m))
    rownames(binNET.m) <- rownames(zNET.m)
    colnames(binNET.m) <- colnames(zNET.m)
   }

  
  
  
  
  ### number of targets per tf
  ntgTF.v <- apply(binNET.m, 2, sum)
  ### number of regulators per gene
  nregG.v <- apply(binNET.m, 1, sum)
  
  

  ### select TFs with at least minNtgts targets
  selTF.idx <- which(ntgTF.v >= minNtgts)
  selbinNET.m <- binNET.m[, selTF.idx]
  selpvNET.m <- pvNET.m[, selTF.idx]
  selzNET.m <- zNET.m[, selTF.idx]
  selcorNET.m <- corNET.m[, selTF.idx]

  mapTG.idx <- match(rownames(selbinNET.m), rownames(exp.m))
  mapTF.idx <- match(colnames(selbinNET.m), rownames(exp.m))

  idx.l <- as.list(seq_len(nrow(selbinNET.m)))
  #print("Computing Partial Correlations")
  pcor.l <- mclapply(idx.l, ComputePCOR, mapTG.idx, mapTF.idx, selbinNET.m, exp.m, mc.cores = ncores)

  pcorNET.m <- matrix(0, nrow = nrow(selbinNET.m), ncol = ncol(selbinNET.m))
  rownames(pcorNET.m) <- rownames(selbinNET.m)
  colnames(pcorNET.m) <- colnames(selbinNET.m)

  #<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< 
  for (g in seq_len(length(pcor.l))) {
    reg.idx <- which(selbinNET.m[g,] == 1)
    if (length(reg.idx) >= 2) {
      pcorNET.m[g, reg.idx] <- pcor.l[[g]][1, -1]
    }
    else if (length(reg.idx) == 1) {
      pcorNET.m[g, reg.idx] <- selcorNET.m[g, reg.idx]
    }
  }
 #>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

  gtexNET.m <- sign(pcorNET.m)
  gtexNET.m[abs(pcorNET.m) < pcorth] <- 0

  sumnetTF.m <- matrix(nrow = 4, ncol = ncol(gtexNET.m))
  rownames(sumnetTF.m) <- c("nTG", "nUP", "nDN", "P")
  colnames(sumnetTF.m) <- colnames(gtexNET.m)
  sumnetTF.m[1, ] <- apply(abs(gtexNET.m), 2, sum)

  #<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
  for(tf in 1:ncol(gtexNET.m)){
    sumnetTF.m[2,tf] <- length(which(gtexNET.m[,tf]==1));
    sumnetTF.m[3,tf] <- length(which(gtexNET.m[,tf]==-1));    
  }
  
  
  #>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

  pv.v <- pbinom(apply(sumnetTF.m[2:3, ], 2, max), size = sumnetTF.m[1, ], prob = 0.5, lower.tail = FALSE)
  sumnetTF.m[4, ] <- pv.v

  gtexNETf.m <- gtexNET.m[, which(sumnetTF.m[1, ] >= minNtgts)]
  
  
  ### now which TFs are overexpressed in toi
  topTOI.lm <- list()
  ### compare toi to all other tissues
  pheno.v <- rep(0, ncol(exp.m))
  pheno.v[which(tissue == toi)] <- 1
  lim.o <- LimmaFn(pheno.v, exp.m)
  topTOI.lm[[1]] <- lim.o$top[[1]]
  ### now compare toi to other tissues (in order to avoid confounding by immune or stromal cell infiltrates)
  
  
  
  if(!is.null(cft)){
    
   
      sel.idx <- which(tissue %in% c(toi,cft));
      tmp.v <- tissue[sel.idx];
      tmpPH.v <- rep(0,length(sel.idx));
      tmpPH.v[which(tmp.v==toi)] <- 1;
      lim.o <- LimmaFn(tmpPH.v,exp.m[,sel.idx]);
      topTOI.lm[[2]] <- lim.o$top[[1]];
     
    
  }
  

  ### now find tissue-specific TFs
  toiTF.lv <- list();
  for(i in 1:length(topTOI.lm)){
    statTF.m <- topTOI.lm[[i]][match(colnames(gtexNETf.m),rownames(topTOI.lm[[i]])),c(1,3,4,5)];
    toiTF.idx <- intersect(which(statTF.m[,4] < degth[i]),which(statTF.m[,1]>lfcth[i]));
    toiTF.lv[[i]] <- rownames(statTF.m[toiTF.idx,]);
  }
  
 # toiTF.v <- toiTF.lv[[1]];
 #if(length(toiTF.lv)>1){
 # for(i in 2:length(toiTF.lv)){
 #  toiTF.v <- intersect(toiTF.v,toiTF.lv[[i]]);
 #}
 #}
  
  
   toiTF.v <- toiTF.lv[[1]];
  if(length(toiTF.lv)>1){
   for(i in 2:length(toiTF.lv)){
    toiTF.v <- intersect(toiTF.v,toiTF.lv[[i]]);
  }
  }
  
  
  summ <- summary(factor(unlist(lapply(toiTF.lv, unique))))
  toiTF.v <- names(summ)[summ == length(toiTF.lv)]
  
  
  map.idx <- match(toiTF.v, colnames(gtexNETf.m))

  if (length(map.idx) <= 1) stop(paste0("Only ", length(map.idx), "TFs in your network. Could try releasing the threshold 'lfcth'/'degth'/'minNtgts' to keep more."))

  #### tissue-specific regulatory network is:
  netTOI.m <- gtexNETf.m[, map.idx]

  distNet.m <- matrix(nrow = ncol(netTOI.m), ncol = 3)
  rownames(distNet.m) <- colnames(netTOI.m)
  colnames(distNet.m) <- c("nTGTS", "Act", "Rep")
  distNet.m[, 1] <- apply(abs(netTOI.m), 2, sum)
  
  for(c in 1:ncol(netTOI.m)){
    act.idx <- which(netTOI.m[,c]==1)
    inact.idx <- which(netTOI.m[,c]==-1)
    distNet.m[c,2:3] <- c(length(act.idx),length(inact.idx));
  }
  

  return(list(netTOI = netTOI.m, sumnet = distNet.m, top = topTOI.lm))
}  ### end of function sciraInfNet
